timeutil/schedule.go

```diff
@@ -0,0 +1,252 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package timeutil
+
+import (
+	"fmt"
+	"math/rand"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var validTime = regexp.MustCompile(`^([0-9]|0[0-9]|1[0-9]|2[0-3]):([0-5][0-9]):?([0-5][0-9])?$`)
+
+type TimeOfDay struct {
+	Hour   int
+	Minute int
+	Second int
+}
+
+func (td TimeOfDay) Less(other TimeOfDay) bool {
+	if td.Hour < other.Hour {
+		return true
+	}
+	if td.Minute < other.Minute {
+		return true
+	}
+	if td.Second < other.Second {
+		return true
+	}
+	return false
+}
+
+// ParseTime parses a string that contains hour:minute and returns
+// an TimeOfDay type or an error
+func ParseTime(s string) (t TimeOfDay, err error) {
+	m := validTime.FindStringSubmatch(s)
+	if len(m) == 0 {
+		return t, fmt.Errorf("cannot parse %q", s)
+	}
+	t.Hour, err = strconv.Atoi(m[1])
+	if err != nil {
+		return t, fmt.Errorf("cannot parse %q: %s", m[1], err)
+	}
+	t.Minute, err = strconv.Atoi(m[2])
+	if err != nil {
+		return t, fmt.Errorf("cannot parse %q: %s", m[2], err)
+	}
+	if m[3] != "" {
+		t.Second, err = strconv.Atoi(m[3])
+		if err != nil {
+			return t, fmt.Errorf("cannot parse %q: %s", m[3], err)
+		}
+	}
+	return t, nil
+}
+
+// Schedule defines a start and end time and an optional weekday in which
+// events should run.
+type Schedule struct {
+	Start TimeOfDay
+	End   TimeOfDay
+
+	Weekday string
+}
+
+func (sched *Schedule) Next(last time.Time) (start, end time.Time) {
+	now := timeNow()
+	wd := time.Weekday(weekdayMap[sched.Weekday])
+
+	t := last
+	for {
+		a := time.Date(t.Year(), t.Month(), t.Day(), sched.Start.Hour, sched.Start.Minute, sched.Start.Second, 0, time.Local)
+		b := time.Date(t.Year(), t.Month(), t.Day(), sched.End.Hour, sched.End.Minute, sched.End.Second, 0, time.Local)
+
+		// not using AddDate() here as this can panic() if no
+		// location is set
+		t = t.Add(24 * time.Hour)
+
+		// we have not hit the right day yet
+		if sched.Weekday != "" && a.Weekday() != wd {
+			continue
+		}
+		// same inteval as last update, move forward
+		if last.After(a) && last.Before(b) {
+			continue
+		}
+		// schedule is right now
+		if now.After(a) && now.Before(b) {
```

Comment:
The logic here is still not handling edges correctly (`a == now` or `b == now`).

Any reason to not take the previous suggestion here, given I think it's both more correct and simpler? Specifically, instead of this branch and all the logic below, using this:
```
if b.Before(now) {
        continue
}
return a, b
```

Per previous explanation, if `b >= now`, there are only two possibilities:

1. `a <= now` - we're inside the interval
2. `a > now` - it's the next interval

In both cases, we want to `return a, b`

If that's off for some reason feel free to ignore me tomorrow, but I _think_ that'd work better with less code.